be/src/exec/pipeline/nljoin/spillable_nljoin_probe_operator.cpp

```diff
@@ -0,0 +1,255 @@
+        SlotDescriptor* slot = _col_types[i];
+        ColumnPtr& dst_col = dst->get_column_by_slot_id(slot->id());
+        if (is_probe) {
+            ColumnPtr& src_col = _probe_chunk->get_column_by_slot_id(slot->id());
+            dst_col->append_value_multiple_times(*src_col, _probe_row_current, cur_build_chunk_rows, true);
+        } else {
+            ColumnPtr& src_col = build_chunk->get_column_by_slot_id(slot->id());
+            dst_col->append_shallow_copy(*src_col);
+        }
+    }
+}
+
+SpillableNLJoinProbeOperator::SpillableNLJoinProbeOperator(
+        OperatorFactory* factory, int32_t id, int32_t plan_node_id, int32_t driver_sequence, TJoinOp::type join_op,
+        const std::string& sql_join_conjuncts, const std::vector<ExprContext*>& join_conjuncts,
+        const std::vector<ExprContext*>& conjunct_ctxs, const std::vector<SlotDescriptor*>& col_types,
+        size_t probe_column_count, const std::shared_ptr<NLJoinContext>& cross_join_context)
+        : OperatorWithDependency(factory, id, "spillable_nestloop_join_probe", plan_node_id, driver_sequence),
+          _prober(join_op, join_conjuncts, conjunct_ctxs, col_types, probe_column_count),
+          _cross_join_context(cross_join_context) {}
+
+Status SpillableNLJoinProbeOperator::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(Operator::prepare(state));
+    _accumulator.set_desired_size(state->chunk_size());
+    RETURN_IF_ERROR(_prober.prepare(state, _unique_metrics.get()));
+    _spill_factory = std::make_shared<spill::SpillerFactory>();
+    _spiller = _spill_factory->create({});
+    _cross_join_context->incr_prober();
+    return Status::OK();
+}
+
+void SpillableNLJoinProbeOperator::close(RuntimeState* state) {
+    _cross_join_context->decr_prober(state);
+    Operator::close(state);
+}
+
+bool SpillableNLJoinProbeOperator::is_ready() const {
+    auto res = _cross_join_context->is_right_finished();
+    if (res) {
+        _init_chunk_stream();
+    }
+    return _cross_join_context->is_right_finished();
+}
+
+bool SpillableNLJoinProbeOperator::is_finished() const {
+    return _is_finished || (_is_finishing && _prober.probe_finished() && _build_finished());
+}
+
+bool SpillableNLJoinProbeOperator::has_output() const {
+    return _chunk_stream && _chunk_stream->has_output();
+}
+
+bool SpillableNLJoinProbeOperator::need_input() const {
+    return _prober.probe_finished() && _build_finished();
+}
+
+Status SpillableNLJoinProbeOperator::set_finishing(RuntimeState* state) {
+    // set finishing
+    _is_finishing = true;
+    return Status::OK();
+}
+
+Status SpillableNLJoinProbeOperator::set_finished(RuntimeState* state) {
+    _is_finished = true;
+    return _cross_join_context->finish_one_left_prober(state);
+}
+
+StatusOr<ChunkPtr> SpillableNLJoinProbeOperator::pull_chunk(RuntimeState* state) {
+    TRACE_SPILL_LOG << "pull_chunk:" << _driver_sequence;
+    if (_prober.probe_finished() || _build_chunk == nullptr || _build_chunk->is_empty()) {
```

Comment:
here use probe_finshed() as if condition, does it means right outer/semi/anti or full join needs process remaining rows of the build side?

Reply:
add some comment.

Reply:
now spillable nl join probe operator only support cross join